### ISSUE TITLE
RpcModule::register_blocking_method with panicking method

### ISSUE THREAD

**Issue by koivunej** [paritytech/jsonrpsee#794]
I noticed that jsonrpsee-0.11 and 0.13.x don't handle gracefully `RpcModule::register_blocking_method` methods which panic. Reproducer can be found at https://gist.github.com/koivunej/8eeb6cf7cd5cfcae3696eb458a10297e

I was initially thinking that this is because the response object is built within the `spawn_blocking` the actual callback is executed: https://docs.rs/jsonrpsee-core/0.13.1/src/jsonrpsee_core/server/rpc_module.rs.html#595-612 However running the reproducer with stable 1.61.0 gives an additional expect failing:

```
thread 'tokio-runtime-worker' panicked at 'not yet implemented: this causes a joinerror', src/main.rs:22:17
thread 'tokio-runtime-worker' panicked at 'Sender is still alive managed by us above; qed', /.../.cargo/registry/src/github.com-1ecc6299db9ec823/jsonrpsee-http-server-0.13.1/src/server.rs:712:25
```

Which is here: https://github.com/paritytech/jsonrpsee/blob/4706aadbb08b86f33efbf35442a080cdd478d735/http-server/src/server.rs#L712

I think an expected graceful handling would be similar what we are doing right now in pathfinder, which is almost (in methods registered as async methods):

```rust
let jh = tokio::task::spawn_blocking(|| { /* actual implementation */ });
jh.await
  .context("Database read panic or shutting down") // from anyhow::Context
  .map_err(|e| {
    use jsonrpsee::types::{error::CallError, ErrorObject};
    
    Error::Call(CallError::Custom(ErrorObject::owned(
        jsonrpsee::types::error::ErrorCode::InternalError.code(),
        format!("{}: {}", jsonrpsee::types::error::INTERNAL_ERROR_MSG, e),
        None::<()>,
    )))
  })
  .and_then(|x| x) // flatten is unstable
```

We would like to use `register_blocking_method` in cases where we only have blocking code to execute. We could add a layer of `catch_unwind` to use `register_blocking_method` but that would take up as many (or more) lines as the current spawn_blocking usage from async context.

I haven't checked if the `register_async_method` guard against panicking, but that is less of a concern for us.

**Comment by niklasad1** [comment]
Thanks for reporting.

On our side `panicking` in the callback is not the same as a `JoinError` as we execute this as a async fn which waits on the result from `spawn_blocking` and if the callback panics then the entire thread will panic.

So indeed, you need to return an `Error` from the callback for that to be gracefully handled that is how it's designed.

The same applies for `async methods` at least for the HTTP server.

**Comment by niklasad1** [comment]
I guess we could send an error from the HTTP server if this occurs, the hyper service should be spawned on hyper and if that callback fails i.e, no response is produced.

Thus, then no message is sent to the channel then we could just return an error instead.

**Comment by niklasad1** [comment]
This is fixed by #793
